yt/chyt/controller/internal/agent/agent.go

```diff
@@ -463,10 +483,63 @@ func (a *Agent) Start() {
 		a.OperationNamespace(),
 		a.l)
 
+	a.opletBatchCh = make(chan map[string]*strawberry.Oplet)
+	a.scalingTargetCh = make(chan scalingRequest)
+	go a.runScaler()
+
 	go a.background()
 	a.l.Info("agent started")
 }
 
+func (a *Agent) runScaler() {
+	scalePeriod := time.Duration(a.config.ScalePeriodOrDefault())
+	ticker := time.NewTicker(scalePeriod)
+
+	for {
+		select {
+		case <-a.ctx.Done():
+			return
+		case <-ticker.C:
+			startedAt := time.Now()
+
+			a.l.Info("Starting scaler")
+
+			workerNumber := a.config.ScaleWorkerNumberOrDefault()
+			var wg sync.WaitGroup
+			wg.Add(workerNumber)
+
+			oplets := <-a.opletBatchCh
```

Comment:
Here there is the same race condition. You may access one oplet from both background and runScaler goroutines.
It is OK for fields like `Alias`, but you also access `Active()`, which surely is a race condition.

I think you need to copy all needed fields to a special struct in the background goroutine (you need only `opID` and probably `alias` I guess) and pass this to controller.GetScalerTarget(...) here instead of the oplet

Reply:
I guess the worst thing that could happen is an attempt to check a disabled oplet, but you are right, I don't need the whole oplet, I'll make another struct for it